File: mssp/archaeology/008-cpython-logging-warn/src/DMS/report.py
"""Which channels agreed, which differed, and whether the difference was declared."""


def render(alias, observer, result, policy_note):
    lines = ["", f"== {alias['old_name']} -> {alias['replacement']}  (observer {alias['equivalence']['observer']})"]
    lines.append(f"   host: {alias['host_constraint']}")
    lines.append("")
    for channel, ok, why, values in result["clauses"]:
        lines.append(f"  {'ok ' if ok else '!! '} {channel:<9} {why}")
        if values:
            lines.append(f"        old: {values[0]!r}")
            lines.append(f"        new: {values[1]!r}")
    lines.append("")
    if result["unexercised_permissions"]:
        lines.append(f"  permissions granted but never needed on this run: "
                     f"{', '.join(result['unexercised_permissions'])}")
    else:
        lines.append("  every declared permission was exercised by this run")
    lines.append(f"  {policy_note}")
    lines.append("")
    lines.append(f"  the declaration {'HOLDS' if result['holds'] else 'DOES NOT HOLD'} under this observer,")
    lines.append("  and 'this observer' is three named channels a person chose.")
    return "\n".join(lines) + "\n"

File: mssp/examples/008-compatibility-alias/src/DMS/report.js
// Which declarations hold, which drifted, and what a reader can conclude.

export function render(results, currentVersion) {
  const lines = ["", `== declared compatibility aliases  (current version ${currentVersion})`];
  for (const result of results) {
    const { alias } = result;
    lines.push("");
    lines.push(`  ${result.holds ? "ok " : "!! "} ${alias.old_name} -> ${alias.replacement}` +
               `   window ${alias.valid_from}..${alias.sunset}`);
    lines.push(`      observer ${alias.equivalence.observer}, ` +
               `allowed deltas: ${alias.equivalence.allowed_deltas.join(", ") || "none"}`);
    for (const delta of result.deltas) {
      const mark = delta.allowed ? "permitted" : "NOT PERMITTED";
      lines.push(`      ${mark.padEnd(14)} ${delta.field}: ${JSON.stringify(delta.old)} -> ${JSON.stringify(delta.new)}`);
    }
    for (const problem of result.problems) lines.push(`      PROBLEM  ${problem}`);
  }

  const broken = results.filter((r) => !r.holds);
  lines.push("");
  lines.push(`  declared ${results.length}, holding ${results.length - broken.length}, ` +
             `broken ${broken.length}${broken.length ? `: ${broken.map((r) => r.alias.old_name).join(", ")}` : ""}`);
  lines.push("");
  lines.push("  a declaration is not evidence. Every line above is the run disagreeing or agreeing");
  lines.push("  with something a person wrote down, under an observer that person also chose.");
  return lines.join("\n") + "\n";
}
